### PR TITLE
Add Source §4 GPU

### PR DESCRIPTION
**Pull request by rrtheonlyone**
## Change: Add GPU Acceleration into Source

This PR adds a new variant into the source language (Source §4 GPU) that is able to accelerate source programs using the GPU. 

Demo: https://source-gpu.herokuapp.com/playground

### Documentation

Language specification are documented both in `/docs/md` and latex version is available in `/docs/specs`

### Testing

Both positive and negative tests have been written for both the transpilation process and the runtime checks. Coveralls report shows that we have broken even (lines added to percentage covered). Of course, there is potential for more testing to be done.

### Functionality

All functionality is encompassed in the `gpu` folder of the code. We have made sure not to touch anything outside to allow for multiple features to be added concurrently to the project. However, due to the nature of the project, some changes have been done to the transpiler / astCreator, namely:
- There is a `ignoreWalker` added for certain functions to be ignored by the transpiler (we don't want our GPU kernel function to be transpiled away)
- More creator functions are added in astCreator for use by the GPU transformer

### Integration

Another small PR will be raised to Cadet-Frontend for the integration of the new variant

### CODE REVIEW

File: src/gpu/__tests__/transpile.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    }
+    `
+  const context = mockContext(4, 'gpu')
+  const transpiled = transpile(parse(code, context)!, context.contextId, false, context.variant)
+    .transpiled
+  const cnt = transpiled.match(/__createKernel/g)?.length
+  expect(cnt).toEqual(3)
+})
+
+test('3 for loop case with 1 indices being written to gets transpiled correctly', () => {
```
Comment (yuntongzhang) [inline src/gpu/__tests__/transpile.ts:210]: nitpicking: 'indices' should be 'index'

File: src/gpu/lib.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+      return false
+    }
+  }
+
+  let cnt = 1
+  for (const i of end) {
+    cnt = cnt * i
+  }
+
+  return cnt > 100
```
Comment (yuntongzhang) [inline src/gpu/lib.ts:98]: Shall we make the threshold a constant in some file?

File: src/gpu/transfomer.ts [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+      if (this.innerBody.body.length > 1) {
+        break
+      }
+
+      currForLoop = this.innerBody.body[0]
+    }
+  }
+
+  /*
+   * Based on state, gets the correct body to me run across threads
```
Comment (yuntongzhang) [inline src/gpu/transfomer.ts:267]: nitpicking: looks like 'me' should be 'be'

File: src/transpiler/transpiler.ts [hunk 4]
```diff
@@ -741,6 +798,27 @@ export function transpile(
...
   }
   const functionsToStringMap = generateFunctionsToStringMap(program)
+
+  const gpuDisplayStatements: es.Statement[] = []
+  if (variant === 'gpu') {
+    const res = transpileToGPU(program)
+
+    // add some display statements to program
+    if (res.length > 0) {
+      for (const arr of res) {
```
Comment (yuntongzhang) [inline src/transpiler/transpiler.ts:808]: Would it be better to move the details of generating display statements to somewhere in the `gpu` folder? I'm not sure.

File: src/gpu/__tests__/transpile.ts (hunk 1 above)
Comment (rrtheonlyone) [reply src/gpu/__tests__/transpile.ts:210]: done updated!

File: src/gpu/lib.ts (hunk 2 above)
Comment (rrtheonlyone) [reply src/gpu/lib.ts:98]: yup that's a good idea! We will put it on top of this file for now and update it again later

File: src/gpu/transfomer.ts (hunk 3 above)
Comment (rrtheonlyone) [reply src/gpu/transfomer.ts:267]: yup that's right!

File: src/transpiler/transpiler.ts (hunk 4 above)
Comment (rrtheonlyone) [reply src/transpiler/transpiler.ts:808]: I see what you mean, yes let's move it to `\gpu` folder : all GPU related logic should be housed there.

**Comment by martin-henz** [comment]
A bit hesitant for this to be a silent optimization @openorclose, for didactic reasons. I would like to introduce it explicitly in the course, with an explanation of how to write programs for the optimization to apply.

File: scripts/jsdoc.sh [hunk 5]
```diff
@@ -136,6 +136,20 @@ run() {
...
+	     -c docs/jsdoc/conf.json \
+	     -R ${MD}/README_4_GPU.md \
+	     -d ${DST}/"source_4_gpu"/ \
+	     ${LIB}/misc.js \
+	     ${LIB}/math.js \
+		 ${LIB}/list.js \
+	     ${LIB}/stream.js \
+	     ${LIB}/array.js \
+	     ${LIB}/pairmutator.js \
+	     ${LIB}/concurrency.js
```
[the remark this answers is not in the document]
Comment (rrtheonlyone) [reply scripts/jsdoc.sh:152]: Okay will update!

**Pull request merged**